domain/keyupdater/state/state.go

```diff
@@ -0,0 +1,111 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/core/database"
+	coremachine "github.com/juju/juju/core/machine"
+	"github.com/juju/juju/domain"
+	machineerrors "github.com/juju/juju/domain/machine/errors"
+)
+
+type State struct {
+	*domain.StateBase
+}
+
+// AllPublicKeysQuery returns a state SQL query for fetching the public keys
+// available on a model. This is useful for constructing authorised keys
+// watchers.
+func (s *State) AllPublicKeysQuery() string {
+	return "SELECT public_key FROM user_public_ssh_key"
+}
+
+// AuthorisedKeysForMachine returns a list of authorised public ssh keys for a
+// machine name. If no machine exists for the given machine name an error
+// satisfying [machineerrors.NotFound] will be returned.
+func (s *State) AuthorisedKeysForMachine(
+	ctx context.Context,
+	machineName coremachine.Name,
+) ([]string, error) {
+	db, err := s.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	machineArg := machineID{machineName.String()}
+	machineStmt, err := s.Prepare(`
+SELECT machine_id AS &machineID.*
+FROM machine
+WHERE machine_id = $machineID.machine_id
+`, machineArg)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"preparing select statement for getting machine %q when determining authorised keys: %w",
+			machineName, err,
+		)
+	}
+
+	stmt, err := s.Prepare(`
+SELECT public_key AS &authorisedKey.*
+FROM user_public_ssh_key
+`, authorisedKey{})
+	if err != nil {
+		return nil, fmt.Errorf(
+			"preparing select statement for getting machine %q authorised keys: %w",
+			machineName, err,
+		)
+	}
+
+	authorisedKeys := []authorisedKey{}
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// Because we have two queries to run and two error paths we need to
+		// handle the errors inside this TX so we can produce the correct error.
+		err := tx.Query(ctx, machineStmt, machineArg).Get(&machineArg)
```

Comment:
Not quite sure why we have two transactions here, I can't see where the return value of this first transaction is being used.

And if we do need both, is it possible to combine them into one statement/txn using `JOIN` or the like?

Reply:
There is only one transaction here. What you mean is why is there two database queries. The tables being queried have no common references at the moment so we can't join them like that. It is cheaper to perform the first query as a check first and then the second one. If we change how authorised keys changes over time to have referential integrity then it may make sense to perform this.

The great thing about the design is this can change under the hood over time.